lib/fileset/internal.nix

```diff
@@ -266,9 +325,92 @@ rec {
     in
     # Special case because the code below assumes that the _internalBase is always included in the result
     # which shouldn't be done when we have no files at all in the base
+    # This also forces the tree before returning the filter, leads to earlier error messages
     if tree == null then
       empty
     else
       nonEmpty;
 
+  # Computes the union of a list of filesets.
+  # The filesets must already be coerced and validated to be in the same filesystem root
+  # Type: [ Fileset ] -> Fileset
+  _unionMany = filesets:
+    let
+      first = head filesets;
+
+      # To be able to union filesetTree's together, they need to have the same base path.
+      # Base paths can be unioned by taking their common prefix,
+      # e.g. such that `union /foo/bar /foo/baz` has the base path `/foo`
+
+      # A list of path components common to all base paths
+      commonBaseComponents = foldl'
```

Comment:
The `foldl'` will only evaluate to WHNF for each item, which means that it can still build thunks (in e.g. an attr) that require a proportional amount of stack space when they are forced _later_.

Could you add a test with a massive number of subdirectories? I don't know which kind of structure is most risky. Deep directories for sure, but those are probably not relevant. Maybe something like `./src/dir<n>/file`.

The evaluator is responsive to `ulimit -s`, so you could perhaps use that to encounter problems sooner, and figure out numbers that make sense to avoid false positives and false negatives with enough margin for Nix to change its stack consumption proportions.